libsolidity/ast/Types.h

```diff
@@ -208,14 +208,30 @@ class Type
 	virtual bool operator==(Type const& _other) const { return category() == _other.category(); }
 	virtual bool operator!=(Type const& _other) const { return !this->operator ==(_other); }
 
-	/// @returns number of bytes used by this type when encoded for CALL. If it is a dynamic type,
-	/// returns the size of the pointer (usually 32). Returns 0 if the type cannot be encoded
-	/// in calldata.
+	/// @returns number of bytes used by this type when encoded for CALL. Cannot be used for
+	/// dynamically encoded types.
+	/// Always returns a value greater than zero and throws if the type cannot be encoded in calldata
+	/// (or is dynamically encoded).
 	/// If @a _padded then it is assumed that each element is padded to a multiple of 32 bytes.
-	virtual unsigned calldataEncodedSize(bool _padded) const { (void)_padded; return 0; }
+	virtual unsigned calldataEncodedSize(bool _padded) const { (void)_padded; solAssert(false, ""); }
+	/// @returns the distance between two elements of this type in an array, tuple or struct.
+	/// For statically-sized types this is the same as calldataEncodedSize(true).
+	/// For dynamically-encoded types this is the distance between two tail pointers, i.e. 32.
+	/// Always returns a value greater than zero and throws if the type cannot be encoded in calldata.
+	virtual unsigned calldataHeadIncrement() const { return calldataEncodedSize(); }
+	/// @returns the (minimal) size of the calldata tail for this type. Can only be used for
+	/// dynamically encoded types. For dynamically-sized arrays this is 32 (the size of the length),
+	/// for statically-sized, but dynamically-encoded arrays this is 32*static_length, for structs
```

Comment:
Is it the size of the tail including the sub-tails?

Reply:
So basically for `uint[][10][][10]` is it 10*32*10*32 or just 10*32?

Reply:
No, only one level of tails... it's spelled out explicitly for all cases, so that should explain it... Any suggestion to improve the first sentence to make it clear right away?

Reply:
For ``uint[][10][][10]`` it'd be ``10*32``. Not sure what I'd do with the value ``10*32*10*32``...

Reply:
Ah, but that's mainly because the dynamic array can be empty. What about `uint[][10][10]`?

Reply:
Will changing
```
for statically-sized, but dynamically-encoded arrays this is 32*static_length
```
to
```
for statically-sized, but dynamically-encoded arrays this is 32*length()
```
help? I'll do that in any case.

Reply:
Yeah, I think so.